File: c/homework/2jintree.c
#include <stdio.h>
#include <stdlib.h>

int dan = 0,dd = 0;
// 이진트리의 노드를 표현하는 구조체 정의
typedef struct TreeNode {
    int data;                // 노드에 저장될 데이터
    struct TreeNode *left;    // 왼쪽 자식 포인터
    struct TreeNode *right;   // 오른쪽 자식 포인터
} TreeNode;


/* ==========================
   노드 개수 계산
   ========================== */
int count_node(TreeNode* root) {
    // 노드가 없으면 0개
    if (root == NULL)
        return 0;
    dan += root->data;
    // 자기 자신(1) + 왼쪽 서브트리 + 오른쪽 서브트리
    return 1 + count_node(root->left) + count_node(root->right);
}

/* ==========================
   단말(leaf) 노드 개수 계산
   ========================== */
int count_leaf(TreeNode* root) {
    // 노드가 없으면 단말도 없음
    if (root == NULL)
        return 0;

    // 왼쪽, 오른쪽 자식이 모두 없으면 단말 노드
    if (root->left == NULL && root->right == NULL)
    {
        dd += root->data;
        return 1;
    }
    // 단말이 아니면 왼쪽 + 오른쪽 결과 합
    return count_leaf(root->left) + count_leaf(root->right);
}

/* ==========================
   트리 높이 계산
   ========================== */
int calc_height(TreeNode* root) {
    // 노드가 없으면 높이 0
    if (root == NULL)
        return 0;

    // 왼쪽과 오른쪽 서브트리 높이 계산
    int left_h = calc_height(root->left);
    int right_h = calc_height(root->right);

    // 더 큰 쪽에 1(현재 노드) 추가
    return (left_h > right_h ? left_h : right_h) + root -> data;
}


int main() {

    TreeNode *a, *b, *c, *d, *e, *f, *g;

    // 노드 메모리 할당
    a= (TreeNode*)malloc(sizeof(TreeNode));
    b = (TreeNode*)malloc(sizeof(TreeNode));
    c = (TreeNode*)malloc(sizeof(TreeNode));
    d= (TreeNode*)malloc(sizeof(TreeNode));
    e = (TreeNode*)malloc(sizeof(TreeNode));
    f = (TreeNode*)malloc(sizeof(TreeNode));
    g = (TreeNode*)malloc(sizeof(TreeNode));

    /*
            A
           / \
          B   C
         / \
        D   E
    */

    // 데이터 저장
    a->data = 10;
    b->data = 20;
    c->data = 30;
    d->data = 40;
    e->data = 50;
    f -> data = 60;
    g -> data = 70;

    // 트리 연결
    a->left = b;
    a->right = c;

    b->left = d;
    b->right = e;

    c->left = f;
    c->right = NULL;

    d->left = NULL;
    d->right = NULL;

    e->left = NULL;
    e->right = NULL;

    f -> left = NULL;
    f -> right = g;
    
    g -> left = NULL;
    g -> left = NULL;

    // ===== 결과 출력 =====
    count_node(a);
    printf("TotalCost : %d\n",dan);
    count_leaf(a);
    printf("FinalCost : %d\n", dd);
    printf("MaxCost   : %d\n", calc_height(a));

    // 메모리 해제 (자식 → 부모 순서)
    free(g);
    free(f);
    free(d);
    free(e);
    free(b);
    free(c);
    free(a);

    return 0;
}

File: python/homework/2jintree.py

# 이진트리 노드 클래스
class TNode:
    def __init__(self, data, left, right):
        self.data = data
        self.left = left
        self.right = right


# ------------------------------
# 노드 개수 계산
# ------------------------------
def count_node(n):
    global dan
    # 종료 조건: 노드가 없으면 0
    if n is None:
        return 0
    dan += n.data
    # 자기 자신(1) + 왼쪽 + 오른쪽
    return 1 + count_node(n.left) + count_node(n.right)


# ------------------------------
# 단말(leaf) 노드 개수 계산
# ------------------------------
def count_leaf(n):
    global dd
    # 노드가 없으면 단말도 없음
    if n is None:
        return 0

    # 왼쪽, 오른쪽이 모두 없으면 단말 노드
    if n.left is None and n.right is None:
        return 1
    dd += n.data
    # 단말이 아니면 왼쪽 + 오른쪽 결과 합
    return count_leaf(n.left) + count_leaf(n.right)


# ------------------------------
# 트리 높이 계산
# ------------------------------
def calc_height(n):
    # 노드가 없으면 높이 0
    if n is None:
        return 0

    # 왼쪽과 오른쪽 높이 계산
    left_h = calc_height(n.left)
    right_h = calc_height(n.right)

    # 더 큰 쪽에 1(현재 노드) 추가
    return max(left_h, right_h) + n.data


if __name__ == "__main__":
    """
            A
           / \
          B   C
         / \
        D   E
    """
    dd = 0
    dan = 0
    # 트리 생성
    g = TNode(70, None, None)
    f = TNode(60,None,g)
    d = TNode(40, None, None)
    e = TNode(50, None, None)
    b = TNode(20, d, e)
    c = TNode(30, f, None)
    a = TNode(10, b, c)

    # 결과 출력
    count_node(a)
    print("노드 개수 :", dan)
    count_leaf(a)
    print("단말 개수 :", dd)
    print("트리 높이 :", calc_height(a))